### PR TITLE
map raster

### PR DESCRIPTION
**Pull request by ArneDefauw**
updated 'apply' function for raster data.

Following discussion https://github.com/scverse/spatialdata/pull/407, I decided to remove the 'combine_c', 'combine_z' parameters, and also support for different Callable's/fn_kwargs per channel/z-stack. I think it overcomplicates the code; the user can simply use a for loop if different Callables/fn_kwargs should be passed to different channels/z-stacks. 

see unit tests for examples:

https://github.com/ArneDefauw/spatialdata/blob/b3e4487cbdfe34ccf38b5a311de1eb2a4b5427c4/tests/core/operations/test_map.py#L31

Signature is now as follows

```
def map_raster(
    data: SpatialImage | MultiscaleSpatialImage,
    func: Callable,
    fn_kwargs: Mapping[str, Any] = MappingProxyType({}),
    chunks: str | int | tuple[int, ...] | tuple[tuple[int, ...], ...] | None = None,
    output_chunks: tuple[tuple[int, ...], ...] | None = None,
    depth: str | int | tuple[int, ...] | dict[int:int] | None = None,
    scale_factors: ScaleFactors_t | None = None,  # if specified will return multiscale
    c_coords: int | str | Iterable[int | str] | None = None, 
    **kwargs,
) -> SpatialImage | MultiscaleSpatialImage:
```

I can not see how we can remove `scale_factors` or `c_coords` from the signature without loosing some functionality.

In order for `map_raster` to be able to alter dimension (e.g. (c,z,y,x)->(c,y,x) ), we would need also need to pass `dims` I think, see https://github.com/ArneDefauw/spatialdata/blob/b3e4487cbdfe34ccf38b5a311de1eb2a4b5427c4/tests/core/operations/test_map.py#L149

### CODE REVIEW

**Comment by LucaMarconato** [comment]
Before the full review, two quick questions. 

1. I see that `chunks` is used only in this line here `arr = da.asarray(arr).rechunk(chunks)`. I would remove the `chunks` argument and have the user rechunk the data before calling `map_raster()`, in this way we have one less argument and the function is more transparent. Or, what's the argument to delegate this task inside `map_raster()`?

**Comment by LucaMarconato** [comment]
2. We need to decide how the output type (in the sense of single-scale vs multi-scale) relates to the input type. We have some options:
	1.	the output type is always the same as the input type: if the user passes a single scale, a single-scale image will be given, if the user passes a multiscale image, then each scale will be processed independently and returned as a multiscale object. This means that if the user is interested in processing only the top scale of a multiscale object, they will need to pass `next(iter(my_multiscale['scale0'].data))`
	2. the output type is always by default a single-scale image, which is computed by the largest scale available.
	3. as in 2, but we have an argument `scale_factors` that compute a multiscale object. I actually would prefer 1 or 2. In fact we should rather have an helper function `to_multiscale()` that calls the model and returns the a multiscale object; in fact constructing multiscale objects from a single scale is a common task that the user should be able to do easily beyond the use cases of this PR.
	
What are your thought on this? I would either go for 1 or 2, and as explained above, I would add an helper function to avoid the need to go for 3.

**Comment by LucaMarconato** [comment]
Another minor comment. I will remove the calls to `rechunk` to address the problem of data not being saved if the chunks as irregular. Let's fix this in a more general setting inside `write()`, as also asked by this user: https://github.com/scverse/spatialdata/issues/581.

**Comment by LucaMarconato** [comment]
> Before the full review, two quick questions.
> 
> I see that chunks is used only in this line here arr = da.asarray(arr).rechunk(chunks). I would remove the chunks argument and have the user rechunk the data before calling map_raster(), in this way we have one less argument and the function is more transparent. Or, what's the argument to delegate this task inside map_raster()?

So, there is one test (`test_map_raster_output_chunks()`) in which the `chunk` argument is required. I kept this argument and renamed it to `input_chunks`. 

Nevertheless, most of the time the argument `input_chunks` is not required as the data is already chunked correctly. I have therefore changed the default behavior. Now, if `input_chunks` is `None`, the data is simply used as it is, without the need to call `rechunk`. Now, to apply the function to the full data and not chunkwise (the old `input_chunks=None` behavior), can be achieved by passing `chunkwise=False` (default is `True`).

**Comment by LucaMarconato** [comment]
I have reviewed the code, fixed the pre-commits, simplified the APIs while maintaining all the use cases of the tests. @giovp can you also review please?

Some minor comments:
- [ ] @ArneDefauw I added 3 TODOs to improve the docstrings by adding some examples for those parameters that allow many different types, such as chunks.
- [ ] Some minor tests for the exceptions should be added, for instance passing a non-raster type
- [ ] We should test also against labels, not just images.

**Comment by LucaMarconato** [comment]
I addressed all the points that I left open; some tests are failing in the CI but not locally, I think they are due to the new support `DataArray`/`DataTree` vs `SpatialImage`/`MultiscaleSpatialImage`. After that is addressed, if @giovp approves I'd merge.

**Comment by LucaMarconato** [comment]
I have added the support for `DataArray`/`DataTree`. Some tests were failing because the raster models were returning a `SpatialImage` instead of a `DataTree` due to a call of `to_spatial_image()`. I have fixed this, and I fixed other minor things regarding the `DataArray`/`DataTree` refactoring. @melonora please be aware of them. If you need them you may consider cherry-picking the relevant code from my next commit until this PR is merged.

**Comment by melonora** [comment]
> 2. We need to decide how the output type (in the sense of single-scale vs multi-scale) relates to the input type. We have some options:
>    
>    1. the output type is always the same as the input type: if the user passes a single scale, a single-scale image will be given, if the user passes a multiscale image, then each scale will be processed independently and returned as a multiscale object. This means that if the user is interested in processing only the top scale of a multiscale object, they will need to pass `next(iter(my_multiscale['scale0'].data))`
>    2. the output type is always by default a single-scale image, which is computed by the largest scale available.
>    3. as in 2, but we have an argument `scale_factors` that compute a multiscale object. I actually would prefer 1 or 2. In fact we should rather have an helper function `to_multiscale()` that calls the model and returns the a multiscale object; in fact constructing multiscale objects from a single scale is a common task that the user should be able to do easily beyond the use cases of this PR.
> 
> What are your thought on this? I would either go for 1 or 2, and as explained above, I would add an helper function to avoid the need to go for 3.
> 
> Update: I have implemented option 2. Happy to discuss.

Think option 2 is indeed best as a user can still pull out a particular scale and use apply should the user want a different level. Also with the output the user can still create a multiscale again.

**Comment by melonora** [comment]
I think this PR will require a notebook tutorial

File: src/spatialdata/models/models.py
Comment (melonora) [inline src/spatialdata/models/models.py:185]: to_spatial_image returns a `DataArray` and does some internal validation so this should be reverted.
```suggestion
        data = to_spatial_image(array_like=data, dims=cls.dims.dims, **kwargs)
```

**Review by melonora — APPROVED** [review]
I see no blockers from my side. Great work!

**Comment by ArneDefauw** [comment]
> 2. We need to decide how the output type (in the sense of single-scale vs multi-scale) relates to the input type. We have some options:
>        
>        1. the output type is always the same as the input type: if the user passes a single scale, a single-scale image will be given, if the user passes a multiscale image, then each scale will be processed independently and returned as a multiscale object. This means that if the user is interested in processing only the top scale of a multiscale object, they will need to pass `next(iter(my_multiscale['scale0'].data))`
>        2. the output type is always by default a single-scale image, which is computed by the largest scale available.
>        3. as in 2, but we have an argument `scale_factors` that compute a multiscale object. I actually would prefer 1 or 2. In fact we should rather have an helper function `to_multiscale()` that calls the model and returns the a multiscale object; in fact constructing multiscale objects from a single scale is a common task that the user should be able to do easily beyond the use cases of this PR.
> 
> 
> What are your thought on this? I would either go for 1 or 2, and as explained above, I would add an helper function to avoid the need to go for 3.
> 
> Update: I have implemented option 2. Happy to discuss.


Yes, indeed, while I see value in being able to provide `scale_factors` as a parameter, I think option 2 is indeed the cleanest solution, as converting to multiscale is a bit outside scope of the function.

**Comment by ArneDefauw** [comment]
Regarding this part of the code, I have three comments:

https://github.com/ArneDefauw/spatialdata/blob/71c46654b5d85742d7ee71e3ee630533f276e339/src/spatialdata/_core/operations/map.py#L90

    if not chunkwise:
        arr = func(arr, **fn_kwargs)
        if output_chunks is not None:
            arr = arr.rechunk(output_chunks)
    else:
        if input_chunks is not None:
            arr = arr.rechunk(input_chunks)


1. The introduction of the `chunkwise` parameter is a great choice, as it is more explicit than the `chunks` parameter I previously used. However, now I would remove the `input_chunks` parameter, as the user can simply rechunk on the SpatialImage or MultiscaleSpatialImage via `.chunk(...)`, e.g.

```
    se = map_raster(
        sdata_blobs[img_layer].chunk(100),
        func=_multiply,
        fn_kwargs=fn_kwargs,
        chunkwise=False,
        c_coords=None,
        depth=None,
    )
```

If we want to keep the `input_chunks` parameter, we should also use it to rechunk when `chunkwise` is set to False, I do not see why we should not do this.

2. If we remove `input_chunks`, we can rename `output_chunks` to `chunks` , and then its name is the same as in `da.map_blocks` and `da.map_overlap`; which prevents confusion for users that are familiar with dask that would pass `chunks` to kwargs.
3. I would not rechunk with `output_chunks` if blockwise is False. I would ignore `output_chunks` if blockwise is False, and state as such in the docstring.

**Comment by ArneDefauw** [comment]
I've added a unit test for altering dimension (c,z,y,x) -> (z,y,x), as it was still commented as a todo

**Comment by LucaMarconato** [comment]
Thanks for the comments, I agree on all the changes, if you could make them it would be great thanks.

**Comment by ArneDefauw** [comment]
> Thanks for the comments, I agree on all the changes, if you could make them it would be great thanks.

Ok, will do!

**Comment by ArneDefauw** [comment]
> > Thanks for the comments, I agree on all the changes, if you could make them it would be great thanks.
> 
> Ok, will do!

Done.  I also changed the name of the parameter `chunkwise` to `blockwise`, as `da.map_overlap` and `da.map_blocks` are just special cases of  `da.blockwise`.

**Review by giovp — CHANGES_REQUESTED** [review]
looks really great @ArneDefauw , some general questions:
this seems a much more slimmed down version of #407 , in particular, it seems that no calls to dask image are really ever done, but that it just target use case of having became functions to be applied to raster data. 
- Does this mean that there will be a separate API/method for calls to dask.image?
- For things like `region_props`, where both a label and an image layer are needed, would there also be a separate API/method or will be added to this function in the future?

Considering answers to this question, I wonder if the name of the function should be renamed?

These are all genuine questions, I think this is definitely ready to get in as is, and it's great, I just wanted to raise once more about the roadmap/use case of similar, yet not exactly the case, functionalities and whether they should/will be added to spatialdata.

File: src/spatialdata/_core/operations/map.py
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/map.py:85]: I think in that case it would not appear in kwargs but as an argument, but better to double check.

File: src/spatialdata/_core/operations/map.py
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/map.py:112]: When `get_all=False` (default), `d` is a `BaseTransformation`. The `assert` is to avoid `mypy` errors.

File: src/spatialdata/_core/operations/map.py
Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:22]: As `da.map_blocks` and `da.map_overlap` also use `func`, it would indeed make sense to rename fn_kwargs to func_kwargs. I will make the changes

File: src/spatialdata/_core/operations/map.py
Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:58]: Now it defaults to the channels in the input `data`, see:
        `c_coords = c_coords if c_coords is not None else get_channels(data)`

I think this makes sense, and feels a bit cleaner than setting it to `range(len(output_coords))`, because if you do not alter number of channels in `func`, you just want to copy channels from input to output I guess.

Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:58]: I agree, I will change the code, and update docstring as follows:
```

    c_coords
        The channel coordinates for the output data. If not provided, the channel coordinates of the input data are
        used. If the callable `func` is expected to change the number of channel coordinates,
        this argument should be provided, otherwise will default to range(len(output_coords)).


```

File: src/spatialdata/_core/operations/map.py
Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:67]: I changed it to:

```
    kwargs
        Additional keyword arguments to pass to :func:`dask.array.map_overlap` or :func:`dask.array.map_blocks`.
        Ignored if `blockwise` is set to `False`.
```

I checked, and the link to the dask documenttion is made.

In the description of the docstring I also added:
```
    Apply a Callable to raster data.

    Applies a Callable (`func`) to raster data. If `blockwise` is set to True,
    distributed processing will be achieved with `dask.array.map_overlap`/`dask.array.map_blocks`,
    otherwise `func` is appplied to the full data.
```

Maybe this is a bit too much. We already explain it in description of `blockwise` parameter.

File: src/spatialdata/_core/operations/map.py
Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:96]: I will remove it, it was to cover use case where user would pass more 'depth' dimensions than available in arr.

File: src/spatialdata/_core/operations/map.py
Comment (ArneDefauw) [reply src/spatialdata/_core/operations/map.py:98]: If `depth` would be passed in `kwargs`, it will appear as an argument, not in `kwargs`, so in that respect I do not see a problem. 
Instead of putting `depth` and `chunks` in `kwargs`, we could indeed also do the following:

`map_func(func, arr, chunks=chunks, depth=depth,  **fn_kwargs, **kwargs, dtype=arr.dtype)`

We could also remove `chunks` and `depth` from the list of parameters altogether, and simply let users pass them via kwargs. In that scenario, I would also remove these sanity checks on `depth` and delegate them to `da.map_overlap`.

Re `coerce_depth`. We have three options:
1. Remove coerce_depth from our code, and let users pass whatever map_overlap can work with, i.e. we delegate conversion to a dict to map_overlap (which will also call coerce_depth).
2. Remove coerce_depth, and only accept e.g. tuples; and add some sanity checks. But then we are a bit too restrictive I think, escpecially for users that are used to work with `map_overlap`and prefer to pass e.g. a dict.
3. Keep the code as is. Having coerce_depth in our code, makes it a bit more explicit as what is exactly passed to map_overlap, which could help debugging later on.

I have a preferance for 3, but not too strong of an opinion on it

**Comment by ArneDefauw** [comment]
> looks really great @ArneDefauw , some general questions: this seems a much more slimmed down version of #407 , in particular, it seems that no calls to dask image are really ever done, but that it just target use case of having became functions to be applied to raster data.
> 
>     * Does this mean that there will be a separate API/method for calls to dask.image?
> 
>     * For things like `region_props`, where both a label and an image layer are needed, would there also be a separate API/method or will be added to this function in the future?
> 
> 
> Considering answers to this question, I wonder if the name of the function should be renamed?
> 
> These are all genuine questions, I think this is definitely ready to get in as is, and it's great, I just wanted to raise once more about the roadmap/use case of similar, yet not exactly the case, functionalities and whether they should/will be added to spatialdata.


Functionality of this PR is pretty similar as https://github.com/scverse/spatialdata/pull/407. The only difference is the removal of support for being able to process different channels/z-stacks with different Callable/fn_kwargs, as it introduces some complexity that is not really necessary, I think the user can simply write a for loop if different Callable/fn_kwargs need to be used, and then concat the DataArray as required. 

Re API/method for calls to dask.image. I would be happy to implement this, if this would be considered  an added value. With the `map_raster` function one can now do the following, e.g.:

```
def _filter(arr, sigma=5):
    from dask_image.ndfilters import gaussian_filter

    arr = arr.squeeze(0)
    return gaussian_filter(arr, sigma)[None, ...]


func_kwargs = {"sigma": 5}
se = map_raster(
    sdata_blobs["blobs_image"].sel(c=[0]),
    func=_filter,
    c_coords=None,
    func_kwargs=func_kwargs,
    blockwise=False,
)
```

Re `region_props`, for me this seem to fall a bit outside scope of this function. I think we could potentially extend `map_raster` with support for:
 image -> labels (segmentation)
multiple labels -> labels (e.g.merging of masks )
multiple images -> images

Extracting `region_props` seems to fall under `aggregate` in my opinion, where the result is a table. 

Anyway, it would be good to have a discussion about the roadmap with respect to `map` and `aggregate`.

**Review by giovp — APPROVED** [review]
looks all good to me, thank you @ArneDefauw !  I've updated the docstrings striving to clarify but would ask you and @LucaMarconato to give it another read cause maybe there is still room for improvement. If they look good, we just need release note and then LGTM!

**Review by giovp — APPROVED** [review]
looks all good to me, thank you @ArneDefauw !  I've updated the docstrings striving to clarify but would ask you and @LucaMarconato to give it another read cause maybe there is still room for improvement. If they look good, we just need release note and then LGTM!

**Comment by LucaMarconato** [comment]
Looks great! I will update the changelog and the docs and then merge! 🚀

**Comment by ArneDefauw** [comment]
looks good! Thanks @giovp  only one comment:
```

    dims
        The dimensions of the output data. If not provided, the dimensions of the input data are used. It must be
        specified if the callable changes the data dimensions, e.g. `('c', 'y', 'x') -> ('y', 'x')`.
```

If going from `('c', 'y', 'x') -> ('y', 'x')`, we go from images to labels; I don't think we planned to support going from images to labels with map_raster? 

In practice, code should work for that use case I think, but need to test, I think we only need to change:

`    if model not in (Labels2DModel, Labels3DModel):`

to

`if 'c' not in dims`:

**Comment by LucaMarconato** [comment]
> I don't think we planned to support going from images to labels with map_raster?

I enabled this with the line `model = get_raster_model_from_data_dims(dims)`; which gets the appropriate model for the output data based on the dimension of the output data. 

Therefore the lines `if model not in (Labels2DModel, Labels3DModel)` and `if 'c' not in dims` are equivalent.

**Review by LucaMarconato — APPROVED** [review]

**Comment by ArneDefauw** [comment]
> > I don't think we planned to support going from images to labels with map_raster?
> 
> I enabled this with the line `model = get_raster_model_from_data_dims(dims)`; which gets the appropriate model for the output data based on the dimension of the output data.
> 
> Therefore the lines `if model not in (Labels2DModel, Labels3DModel)` and `if 'c' not in dims` are equivalent.

But the model is obtained from the input data

https://github.com/ArneDefauw/spatialdata/blob/759fd992b6881a03e890e4082ec6c83f26aa1630/src/spatialdata/_core/operations/map.py#L88

So this check is on input data:

https://github.com/ArneDefauw/spatialdata/blob/759fd992b6881a03e890e4082ec6c83f26aa1630/src/spatialdata/_core/operations/map.py#L114

**Comment by LucaMarconato** [comment]
You are right, the check on `c_coords` should be one the output data. I merged already because I need to merge #594, which was opened against the current PR. Please can you make a patch for fixing that behavior?

**Comment by ArneDefauw** [comment]
> You are right, the check on `c_coords` should be one the output data. I merged already because I need to merge #594, which was opened against the current PR. Please can you make a patch for fixing that behavior?

Yes, will do!

**Pull request merged**
